rust/agama-server/src/agama-web-server.rs

```diff
@@ -39,26 +40,270 @@ struct Cli {
+        if addr.ip().to_canonical().is_loopback() {
+            // accept plain HTTP on the local connection
+            service.clone().call(request)
+        } else {
+            // redirect external connections to HTTPS
+            redirector_service.clone().call(request)
+        }
+    });
+
+    let ret = Builder::new(TokioExecutor::new())
+        .serve_connection_with_upgrades(stream, hyper_service)
+        .await;
+
+    if let Err(err) = ret {
+        tracing::error!("Error serving connection from {}: {}", addr, err);
+    }
+}
+
+/// Starts the web server
+async fn start_server(address: String, service: Router, ssl_acceptor: SslAcceptor) {
+    tracing::info!("Starting Agama web server at {}", address);
+
+    // see https://github.com/tokio-rs/axum/blob/main/examples/low-level-openssl/src/main.rs
+    // how to use axum with openSSL
+    let listener = tokio::net::TcpListener::bind(&address)
+        .await
+        .unwrap_or_else(|error| {
+            let msg = format!("Error: could not listen on {}: {}", &address, error);
+            tracing::error!(msg);
+            panic!("{}", msg)
+        });
+
+    pin_mut!(listener);
+
+    let redirector = https_redirect();
+
+    loop {
+        let tower_service = service.clone();
+        let redirector_service = redirector.clone();
+        let tls_acceptor = ssl_acceptor.clone();
+
+        // Wait for a new tcp connection; if it fails we cannot do much, so print an error and die
+        let (tcp_stream, addr) = listener
+            .accept()
+            .await
+            .expect("Failed to open port for listening");
+
+        tokio::spawn(async move {
+            if is_ssl_stream(&tcp_stream).await {
+                // handle HTTPS connection
+                handle_https_stream(tls_acceptor, addr, tcp_stream, tower_service).await;
+            } else {
+                // handle HTTP connection
+                handle_http_stream(addr, tcp_stream, tower_service, redirector_service).await;
+            }
+        });
+    }
+}
+
 /// Start serving the API.
-///
-/// `args`: command-line arguments.
+/// `options`: command-line arguments.
 async fn serve_command(args: ServeArgs) -> anyhow::Result<()> {
     let journald = tracing_journald::layer().expect("could not connect to journald");
     tracing_subscriber::registry().with(journald).init();
 
-    let listener = tokio::net::TcpListener::bind(&args.address)
-        .await
-        .unwrap_or_else(|_| panic!("could not listen on {}", &args.address));
-
     let (tx, _) = channel(16);
     run_monitor(tx.clone()).await?;
 
     let config = web::ServiceConfig::load()?;
     let dbus = connection_to(&args.dbus_address).await?;
     let service = web::service(config, tx, dbus).await?;
-    axum::serve(listener, service)
-        .await
-        .expect("could not mount app on listener");
+
+    let ssl_acceptor = if let Ok(ssl_acceptor) = args.ssl_acceptor() {
```

Comment:
There is no need to change it now (time pressure), but it is weird that `ServeArgs` is responsible for creating the SSL acceptor.

Reply:
If the server listens on two ports then it would be a good idea to use the same generated self-signed certificate on both ports. So the SSL acceptor needs to be created early.